### PR TITLE
#960 prevent outdated versions from being released.

### PR DESCRIPTION
**Pull request by original-brownbear**
#960 is solved by this:

* Added maven-artifact dependency for its version comparison capabilities
* QnRelease now differentiates 3 cases and answers accordingly
 * no tag given
 * outdated tag
 * tag that is either not a valid version or newer than the most highest valid version tag
* Added class ReleaseTag to encapsulate the version comparison against the repo
* Adjusted and extended QnReleaseTest for the new differentiation between 3 cases
* Added ReleaseTagTest for the new ReleaseTag class

### CODE REVIEW

**Comment by alex-palevsky** [comment]
@original-brownbear Thanks, I will find someone to review this PR soon

**Comment by alex-palevsky** [comment]
@pinaf review this pull request when possible

**Comment by pinaf** [comment]
@alex-palevsky please assign someone else - no time.

**Comment by alex-palevsky** [comment]
> @alex-palevsky please assign someone else - no time.

@pinaf 15 points deducted from your rating

**Comment by alex-palevsky** [comment]
> @alex-palevsky please assign someone else - no time.

@pinaf right, I will find someone else, no problem

**Comment by alex-palevsky** [comment]
@mkordas please review this pull request when possible

**Comment by mkordas** [comment]
@original-brownbear looks impressive, I'll review it today

File: src/main/java/com/rultor/agents/github/qtn/QnRelease.java [hunk 1]
```diff
@@ -65,31 +66,56 @@
...
             issue.repo().coordinates(), issue.number(), comment.number()
         );
-        return new Req.Simple(
-            "release",
-            new ImmutableMap.Builder<String, String>()
-                .put("head_branch", "master")
-                .put(
-                    "head",
+        final Req req;
+        final Matcher matcher = Pattern.compile(".*release.*`(.+)`.*")
```
Comment (mkordas) [inline src/main/java/com/rultor/agents/github/qtn/QnRelease.java]: @original-brownbear isn't it performance problem? shouldn't it be extracted to constant?

File: src/main/java/com/rultor/agents/github/qtn/QnRelease.java [hunk 2]
```diff
@@ -65,31 +66,56 @@
...
+            .matcher(comment.body());
+        if (matcher.matches()) {
```
Comment (mkordas) [inline src/main/java/com/rultor/agents/github/qtn/QnRelease.java]: @original-brownbear instead of `matches()` we could use `find()` and avoid `.*` on both sides

File: src/test/java/com/rultor/agents/github/qtn/ReleaseTagTest.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+ *
+ * @author Armin Braun (me@obrown.io)
+ * @version $Id$
+ * @since 1.62
+ */
+public final class ReleaseTagTest {
+
+    /**
+     * ReleaseTag can deny release for outdated, semantically correct versions.
+     * @throws Exception In case of error.
```
Comment (mkordas) [inline src/test/java/com/rultor/agents/github/qtn/ReleaseTagTest.java]: @original-brownbear see https://github.com/teamed/qulice/issues/733

File: src/main/java/com/rultor/agents/github/qtn/QnRelease.java (hunk 1 above)
Comment (original-brownbear) [inline src/main/java/com/rultor/agents/github/qtn/QnRelease.java]: @mkordas jup 100% agreed, fixing it :)

File: src/test/java/com/rultor/agents/github/qtn/ReleaseTagTest.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+     * ReleaseTag can retrieve the latest release version in the repo.
+     * @throws Exception In case of error.
+     */
+    @Test
+    public void getsReferenceVersion() throws Exception {
+        final Repo repo = new MkGithub().randomRepo();
+        final String latest = "2.2.1";
+        repo.releases().create("1.0");
+        repo.releases().create(latest);
+        repo.releases().create("3.0-beta");
```
Comment (mkordas) [inline src/test/java/com/rultor/agents/github/qtn/ReleaseTagTest.java]: @original-brownbear it is not clear for my why not `3.0-beta` is latest here

Comment (mkordas) [inline src/test/java/com/rultor/agents/github/qtn/ReleaseTagTest.java]: @original-brownbear OK, it is because of `-beta`

File: src/test/java/com/rultor/agents/github/qtn/ReleaseTagTest.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+ * Tests for ${@link ReleaseTag}.
+ *
+ * @author Armin Braun (me@obrown.io)
+ * @version $Id$
+ * @since 1.62
+ */
+public final class ReleaseTagTest {
+
+    /**
+     * ReleaseTag can deny release for outdated, semantically correct versions.
```
Comment (mkordas) [inline src/test/java/com/rultor/agents/github/qtn/ReleaseTagTest.java]: @original-brownbear let's add info to Javadoc about allowed suffixes

File: src/test/java/com/rultor/agents/github/qtn/ReleaseTagTest.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+import com.jcabi.github.Repo;
+import com.jcabi.github.mock.MkGithub;
+import org.hamcrest.MatcherAssert;
+import org.hamcrest.Matchers;
+import org.junit.Test;
+
+/**
+ * Tests for ${@link ReleaseTag}.
+ *
```
Comment (mkordas) [inline src/test/java/com/rultor/agents/github/qtn/ReleaseTagTest.java]: @original-brownbear see https://github.com/teamed/qulice/issues/708, newline not needed

File: src/test/java/com/rultor/agents/github/qtn/QnReleaseTest.java [hunk 7]
```diff
@@ -77,4 +79,48 @@ public void buildsRequest() throws Exception {
...
+            new QnRelease().understand(
+                new Comment.Smart(issue.comments().get(1)), new URI("#")
+            ),
+            Matchers.is(Req.EMPTY)
+        );
+        MatcherAssert.assertThat(
+            new Comment.Smart(issue.comments().get(2)).body(),
+            Matchers.containsString("There is already a release `1.7`")
+        );
+        issue.comments().post("release");
```
Comment (mkordas) [inline src/test/java/com/rultor/agents/github/qtn/QnReleaseTest.java]: @original-brownbear what that does?

File: src/test/java/com/rultor/agents/github/qtn/QnReleaseTest.java [hunk 8]
```diff
@@ -77,4 +79,48 @@ public void buildsRequest() throws Exception {
...
+    }
+
+    /**
+     * QnRelease can deny release when tag name is not given.
```
Comment (mkordas) [inline src/test/java/com/rultor/agents/github/qtn/QnReleaseTest.java]: @original-brownbear is this in scope of this issue?

File: src/test/java/com/rultor/agents/github/qtn/QnReleaseTest.java (hunk 7 above)
Comment (original-brownbear) [inline src/test/java/com/rultor/agents/github/qtn/QnReleaseTest.java]: @mkordas oversight sorry, removing it :)

File: src/main/java/com/rultor/agents/github/qtn/ReleaseTag.java [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+            }
+        }
+        return tag;
+    }
+
+    /**
+     * Checks that a tag is newer than a given reference.
+     * @param reference String
+     * @param tag String
+     * @return Boolean true if tag is new than reference.
```
Comment (mkordas) [inline src/main/java/com/rultor/agents/github/qtn/ReleaseTag.java]: @original-brownbear why not just `True if tag is new than reference` without dot?

File: src/main/java/com/rultor/agents/github/qtn/ReleaseTag.java [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+     */
+    private static boolean newer(final String reference, final String tag) {
+        return new DefaultArtifactVersion(reference).compareTo(
+            new DefaultArtifactVersion(tag)
+        ) < 0;
+    }
+
+    /**
+     * Checks tag string format being a valid release version.
```
Comment (mkordas) [inline src/main/java/com/rultor/agents/github/qtn/ReleaseTag.java]: @original-brownbear I'd propose `Checks that tag is a valid release version`

File: src/main/java/com/rultor/agents/github/qtn/ReleaseTag.java [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+     * @param identifier String tag name.
+     * @return Boolean
```
Comment (mkordas) [inline src/main/java/com/rultor/agents/github/qtn/ReleaseTag.java]: @original-brownbear we need to elaborate more here

File: src/main/java/com/rultor/agents/github/qtn/ReleaseTag.java [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+    ReleaseTag(final Repo rpo, final String version) {
+        this.repo = rpo;
+        this.name = version;
+    }
+
+    /**
+     * Checks if this tag can be released.
+     * A tag can be released if it is either not named as a semantically
+     * correct version or has a higher version number than all existing tags.
+     * @return Boolean
```
Comment (mkordas) [inline src/main/java/com/rultor/agents/github/qtn/ReleaseTag.java]: @original-brownbear put more info, see http://stackoverflow.com/questions/10088311/javadoc-returns-tag-comment-duplication-neccessary

File: src/main/java/com/rultor/agents/github/qtn/ReleaseTag.java [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+ *
+ * @author Armin Braun (me@obrown.io)
+ * @version $Id$
+ * @since 1.62
+ */
+@Immutable
+final class ReleaseTag {
+
+    /**
+     * Pattern matching semantically valid versions.
```
Comment (mkordas) [inline src/main/java/com/rultor/agents/github/qtn/ReleaseTag.java]: @original-brownbear but e.g. `1.0.0-beta` is also valid semver, but patter doesn't match it

File: src/main/java/com/rultor/agents/github/qtn/QnRelease.java (hunk 2 above)
Comment (original-brownbear) [inline src/main/java/com/rultor/agents/github/qtn/QnRelease.java]: @mkordas agreed, changing this.

File: src/test/java/com/rultor/agents/github/qtn/QnReleaseTest.java (hunk 8 above)
Comment (original-brownbear) [inline src/test/java/com/rultor/agents/github/qtn/QnReleaseTest.java]: @mkordas you know the weird way "release questions" are answered at present Rultor does not answer when the backticks with the version inside are missing. I think somewhere down the line an Exception is thrown and then swallowed in this case. There is a ticket about this somewhere and you can also observe the effect of this when you as a non-architect call the release command. It always posts a second "I don't understand you answer", so far me it seemed safest to cover this case in a test and in code here and go from there with the other ticket.
Also it simply seemed the cleanest approach to just handle this case instead of removing the old test, or leaving in this as a strange corner case that no one understands.
To me it makes sense, I'm open to criticism though ? :)

Comment (mkordas) [inline src/test/java/com/rultor/agents/github/qtn/QnReleaseTest.java]: @original-brownbear I see your point, but on the other hand we have this principle... http://www.yegor256.com/2015/06/22/valid-reasons-to-reject-bug-fix.html#it-solves-more-than-one-issue

**Comment by mkordas** [comment]
@original-brownbear I'll finish my review tomorrow

File: src/test/java/com/rultor/agents/github/qtn/QnReleaseTest.java (hunk 8 above)
Comment (original-brownbear) [inline src/test/java/com/rultor/agents/github/qtn/QnReleaseTest.java]: @mkordas I also see your point :), but following that rule strictly we would simply keep the old test and adjust the code to simply handle the absence of a version equal to a valid version tag ?
It seems to me we're sort of acting against the gist of making an understandable fix here, by adding incorrect complication ( and documenting it ) instead of adding less but correct complication.
Let me handle the other things tonight, maybe I'll find a better approach or argument since we were gonna continue tomorrow on this anyways, but you're right for now I think.

**Comment by original-brownbear** [comment]
@mkordas sure, thanks so far!

File: src/test/java/com/rultor/agents/github/qtn/ReleaseTagTest.java (hunk 5 above)
Comment (original-brownbear) [inline src/test/java/com/rultor/agents/github/qtn/ReleaseTagTest.java]: Added more info below this line :)

**Comment by original-brownbear** [comment]
@mkordas you win, reverted the handling of the no tag given case to the old behavior and kept the test around the way it was. This is def a different issue on second thought and does not add any risk or special code here really. Let's handle this elsewhere :) ( I did not hard push anyways so the solution can easily be retrieved from the history by whoever handles this down the line ).
Let me know what I missed :)

File: years.sh [hunk 14]
```diff
@@ -11,6 +12,7 @@ if (grep -r -L "Copyright \+(c) \+2009-.*rultor.com" \
   --include=*.java --include=*.xml --include=*.vm --include=*.groovy \
   --include=*.txt --include=*.fml --include=*.properties} \
   --exclude-dir "target" --exclude-dir "apache-maven-*" . \
+  --exclude-dir ".idea" \
```
Comment (original-brownbear) [inline years.sh]: @mkordas ignore these two lines, it's a Github/Git issue. They are the same as in master but resulted from a 3-way merge in the branch and somehow it doesn't get this right when diffing :)

Comment (mkordas) [inline years.sh]: @original-brownbear the solution would be to remove this commit and push force

File: src/test/java/com/rultor/agents/github/qtn/ReleaseTagTest.java (hunk 5 above)
Comment (mkordas) [inline src/test/java/com/rultor/agents/github/qtn/ReleaseTagTest.java]: @original-brownbear remember about comment addressee :)

Comment (original-brownbear) [inline src/test/java/com/rultor/agents/github/qtn/ReleaseTagTest.java]: @mkordas sorry :), fixed

**Comment by original-brownbear** [comment]
@mkordas you're not waiting for me here are you ? ( just asking not rushing ;) )

**Comment by mkordas** [comment]
@original-brownbear I'm back, thanks for pinging

File: src/main/java/com/rultor/agents/github/qtn/ReleaseTag.java [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+ * @author Armin Braun (me@obrown.io)
+ * @version $Id$
+ * @since 1.62
+ */
+@Immutable
+final class ReleaseTag {
+
+    /**
+     * Pattern matching semantically valid versions, that also only consist in
+     * digits and dots.
```
Comment (mkordas) [inline src/main/java/com/rultor/agents/github/qtn/ReleaseTag.java]: @original-brownbear I'd reword it to `Pattern to match semantically valid versions that consist only of digits and dots.`

**Comment by mkordas** [comment]
@original-brownbear looks good!

**Comment by mkordas** [comment]
@rultor merge

**Comment by original-brownbear** [comment]
@mkordas thanks for the patience here :)

**Comment by original-brownbear** [comment]
@rultor merge :)

**Comment by mkordas** [comment]
@original-brownbear 
> Let's handle this elsewhere :)

Yeah, let's do it, as the code was good.

**Comment by mkordas** [comment]
@original-brownbear do you know the reason of merge failure?

**Comment by original-brownbear** [comment]
@mkordas yes it appears the Gemfile became outdated in the last few hours since the last merge.
I'll fix it sec :)

**Comment by mkordas** [comment]
@original-brownbear see #1037

**Comment by original-brownbear** [comment]
@mkordas See #1038 and resulting #1039, not catching any breaks here. We're stuck until @yegor256 helps us with Rultor PRs I'm afraid.

**Comment by mkordas** [comment]
@original-brownbear right, #1039 is a good one, we need it, otherwise entire development may be stuck for a long time

**Comment by mkordas** [comment]
@yegor256 we need your help in #1038 in order to merge this PR

**Comment by mkordas** [comment]
@rultor merge

**Comment by original-brownbear** [comment]
@rultor merge :)

**Comment by original-brownbear** [comment]
@mkordas o my bad, well nothing bad will happen from this nothing to merge left :) Closing here thanks! :)

**Comment by alex-palevsky** [comment]
@mkordas 10 mins added to @original-brownbear account (our architect), payment ID is `AP-82U72426JU2475832`, since quality is good here

Much obliged! I have added **44 mins** to your account in payment "56dc214415431a1ef2000158", 69 hours and 45 mins spent

review comments (c=29) added as a bonus

+44 to your rating, your total score is [+8231](http://www.netbout.com/b/36080?open=rating)
